### PR TITLE
Correctly keep master in master history if it was active at the beginning of the master history period

### PR DESCRIPTION
**Pull request by masseyke**
This commit fixes a problem with the MasterHistory class. If we have had a single master for a week and then that master went to null 5 seconds ago, we want to say that we have had a non-null master within the last 30 seconds. However before this change, MasterHistory.hasSeenMasterInLastNSeconds(30) would return false in that case because the non-null master had a start time much more than 30 seconds ago. This fix recognizes that a node is master up until the start time of the master that replaces it. This required changing the methods that pruned the master history based on time so that we do not prune masters that are in any way active during the time period.

Here's an example scenario (which is now accounted for in a unit test):
```
node1 (7 days ago)
null (20 seconds ago)
null (10 seconds ago)*
```
The master is clearly null right now, but MasterHistory.hasSeenMasterInLastNSeconds(30) ought to return true because the master was node1 up until 20 seconds ago.

\* I don't think that we would ever set the master to null twice in a row, but the code now accounts for this just in case it happens.

### CODE REVIEW

**Comment by masseyke** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Review by andreidan — COMMENTED** [review]
Thanks for fixing this Keith 

Not sure if this was ready for review, but left some thoughts.

File: server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistory.java [hunk 1]
```diff
@@ -77,7 +76,13 @@ public void clusterChanged(ClusterChangedEvent event) {
             int startIndex = Math.max(0, sizeAfterAddingNewMaster - MAX_HISTORY_SIZE);
             for (int i = startIndex; i < masterHistory.size(); i++) {
                 TimeAndMaster timeAndMaster = masterHistory.get(i);
-                if (timeAndMaster.startTimeMillis >= oldestRelevantHistoryTime) {
+                final long endTime;
```
[the remark this answers is not in the document]
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistory.java:79]: I renamed `endTime` to `currentMasterEndTime`, and put in a few comments.
> Does this check make sense for any _other_ situation than having a list that contains only 2 entries?
This check is used to make sure that we have the list of masters that were actually active for the entire 30 minute period. So take this example where we've got 4 transitions to null:
```
node1 -> null -> node2 -> null -> node1 -> null -> node2 -> null
```
Say that node1 first became master 2 hours ago, and transitioned to null 20 minutes ago, so node1's start time is outside of the range we're looking for. If we only look at start time, or only look at the end time if there are two things in the history list, then what we'd see is:
```
null -> node2 -> null -> node1 -> null -> node2 -> null -> node1
```
So it looks like we've only had 3 transitions to null and we're report a GREEN status.

File: server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistory.java [hunk 2]
```diff
@@ -175,38 +180,65 @@ public static int getNumberOfMasterIdentityChanges(List<DiscoveryNode> masterHis
     }
 
     /**
-     * Returns true if a non-null master was seen at any point in the last nSeconds seconds, or if the last-seen master was more than
-     * nSeconds seconds ago and non-null.
+     * Returns true if a non-null master was seen at any point in the last nSeconds seconds, including if the master was non-null
+     * nSeconds ago but has transitioned to null since then.
```
[the remark this answers is not in the document]
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistory.java:184]: I'll see if I can change the wording to be more clear (I think it's correct but maybe unclear?)

File: server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistory.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (andreidan) [reply server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistory.java]: Ah gotcha, thanks for the details (and for adding code docs ❤️)

**Review by andreidan — APPROVED** [review]
LGTM, thanks for fixing this Keith

**Pull request merged**